Add get_media_upload_limits_for_user and on_media_upload_limit_exceeded callbacks to module API (#18848)
element-hq/synapse: docs/modules/media_repository_callbacks.md

returns `False` will be used. If this happens, Synapse will not call any of the subsequent
implementations of this callback.

### `get_media_upload_limits_for_user`

_First introduced in Synapse v1.137.0_

```python
async def get_media_upload_limits_for_user(user_id: str, size: int) -> Optional[List[MediaUploadLimit]]
```

**<span style="color:red">
Caution: This callback is currently experimental . The method signature or behaviour
may change without notice.
</span>**

Called when processing a request to store content in the media repository.

The arguments passed to this callback are:

* `user_id`: The Matrix user ID of the user (e.g. `@alice:example.com`) making the request.

If the callback returns a list then it will be used as the limits to be applied to the request.

If an empty list is returned then no limits are applied.

If multiple modules implement this callback, they will be considered in order. If a
callback returns `None`, Synapse falls through to the next one. The value of the first
callback that does not return `None` will be used. If this happens, Synapse will not call
any of the subsequent implementations of this callback.

> If no module returns a non-`None` value then the default [media upload limits config](https://element-hq.github.io/synapse/latest/usage/configuration/config_documentation.html#media_upload_limits) will be used.

Reviewer: Relative links work between documentation pages.

```suggestion
If there are no registered modules, or if all modules return `None`, then
the default
[media upload limits config](../usage/configuration/config_documentation.html#media_upload_limits)
will be used.
```

The arguments passed to this callback are:

* `user_id`: The Matrix user ID of the user (e.g. `@alice:example.com`) making the request.

If the callback returns a list then it will be used as the limits to be applied to the request.

If an empty list is returned then no limits are applied.

If multiple modules implement this callback, they will be considered in order. If a
callback returns `None`, Synapse falls through to the next one. The value of the first
callback that does not return `None` will be used. If this happens, Synapse will not call
any of the subsequent implementations of this callback.

If no module returns a non-`None` value then the default [media upload limits config](https://element-hq.github.io/synapse/latest/usage/configuration/config_documentation.html#media_upload_limits) will be used.

### `on_media_upload_limit_exceeded`

_First introduced in Synapse v1.139.0_

```python
async def on_media_upload_limit_exceeded(user_id: str, limit: MediaUploadLimit, sent_bytes: int, attempted_bytes: int) -> None
```

**<span style="color:red">
Caution: This callback is currently experimental . The method signature or behaviour
may change without notice.
</span>**

> Called when a user attempts to upload media that would exceed a configured media upload limit.

Reviewer: Could you include a note stating that this callback will only be called on media workers (that which handles [POST /_matrix/media/v3/upload](https://spec.matrix.org/v1.15/client-server-api/#post_matrixmediav3upload))?

> Called when a user attempts to upload media that would exceed a configured media upload limit.

Reviewer: And a quick one-liner for why a module may find this useful to implement? Perhaps "This can be used to inform the user that they've reached a media upload limit through some external interface".

As it's not too obvious otherwise.

Author: Added in a751ef13d3e0b6d8a9cb0004d247c908bf101a7f

any of the subsequent implementations of this callback.

If no module returns a non-`None` value then the default [media upload limits config](https://element-hq.github.io/synapse/latest/usage/configuration/config_documentation.html#media_upload_limits) will be used.

### `on_media_upload_limit_exceeded`

_First introduced in Synapse v1.137.0_

```python
async def on_media_upload_limit_exceeded(user_id: str, limit: MediaUploadLimit, sent_bytes: int, attempted_bytes: int) -> None
```

**<span style="color:red">

> Caution: This callback is currently experimental . The method signature or behaviour

Reviewer: ```suggestion
Caution: This callback is currently experimental. The method signature or behaviour
```
